ts4.1/index.d.ts

```diff
@@ -158,7 +166,12 @@ type UseTranslationResponse<N extends Namespace> = [TFunction<N>, i18n, boolean]
   ready: boolean;
 };
 
-type DefaultNamespace<T = 'translation'> = T extends Fallback<string> ? T : string;
+type TranslationNS = 'translation';
+type NameKey = 'name';
+
+type ExtractName<T> = T extends { [key in NameKey]: unknown } ? T[NameKey] : TranslationNS;
+
+type DefaultNamespace<T = ExtractName<DefaultNS>> = T extends Fallback<string> ? T : string;
```

Comment:
Hey @JCQuintas, my suggestion would be something like this:
```ts
export interface CustomTypeOptions {}

type MergeBy<T, K> = Omit<T, keyof K> & K;
type TypeOptions = MergeBy<{
  defaultNS: 'translation';
}, CustomTypeOptions>

type DefaultNamespace<T = TypeOptions['defaultNS']> = T extends Fallback<string> ? T : string;
```

Then in the future, we could include resources here as well, even other options that may appear. So instead of having one type augmentation for each option, we could have one for all, for example:

```ts
interface CustomTypeOptions = {
  resources: typeof resources;
  defaulsNS: 'ns1';
}
```

What do you think?

Reply:
I like it. My approach was to not make changes to existing parts of the application cause I don't have a deep understanding of them. 

But I believe your approach to accomplish the same and still be flexible to future needs.

I'll make the changes

Reply:
I've changed it, incorporated the `resources` key as you suggested with backwards compatibility with the `Resources` type.

I've added tests to check if this works, though I had to update the typescript version. I don't think this will be an issue since it is a dev dep, but I don't know if it has other side effects I'm unaware of.

Reply:
If we add resources now, we'll have breaking changes. I'm ok with having breaking changes, but we'd need to hold the PR off for a minor or major release.

//cc: @adrai

Reply:
Never mind, I've just seen you're keeping Resources and adding it to the default one. Just add a `@deprecated` decorator to the old `Resources` interface.

Reply:
> If we add resources now, we'll have breaking changes. I'm ok with having breaking changes, but we'd need to hold the PR off for a minor or major release.

breaking means major

Reply:
Added `@deprecated` with a message to use `CustomTypeOptions` instead. And yes, it should be backwards compatible, so there would be no need for a `major` release

Reply:
@JCQuintas This change did break my team's app. I haven't dug in deep enough to say if it is totally our fault or not - but I can say for sure that:

* Our app will not build with the latest version
* I went through about 6 build errors, hacking in some TS fixes, but the build errors just kept coming.
* I pin the version to "react-i18next": "11.8.6" and the app builds fine

Reply:
For anybody that makes it here with the same problem, we found an easy fix.

![image](https://user-images.githubusercontent.com/14027108/122138073-d7bdf280-cdfa-11eb-9710-1cc12fbf84d8.png)

Reply:
For anybody who makes it here, we found an easy fix: 
![image](https://user-images.githubusercontent.com/14027108/122138138-f9b77500-cdfa-11eb-8e4c-6d266d455282.png)